ledger/voters.go

```diff
@@ -27,87 +27,87 @@ import (
-// will be used to sign the compact cert for certRnd.
-func votersRoundForCertRound(certRnd basics.Round, proto config.ConsensusParams) basics.Round {
-	// To form a compact certificate for round certRnd,
-	// we need a commitment to the voters CompactCertRounds
+// votersRoundForStateProofRound computes the round number whose voting participants
+// will be used to sign the state proof for stateProofRnd.
+func votersRoundForStateProofRound(stateProofRnd basics.Round, proto config.ConsensusParams) basics.Round {
+	// To form a state proof on period that ends on stateProofRnd,
+	// we need a commitment to the voters StateProofInterval rounds
 	// before that, and the voters information from
-	// CompactCertVotersLookback before that.
-	return certRnd.SubSaturate(basics.Round(proto.CompactCertRounds)).SubSaturate(basics.Round(proto.CompactCertVotersLookback))
+	// StateProofVotersLookback before that.
+	return stateProofRnd.SubSaturate(basics.Round(proto.StateProofInterval)).SubSaturate(basics.Round(proto.StateProofVotersLookback))
 }
 
-func (vt *votersTracker) loadFromDisk(l ledgerForTracker, ao *onlineAccounts) error {
+func (vt *votersTracker) loadFromDisk(l ledgerForTracker, fetcher ledgercore.OnlineAccountsFetcher, latestDbRound basics.Round) error {
 	vt.l = l
-	vt.ao = ao
-	vt.round = make(map[basics.Round]*ledgercore.VotersForRound)
+	vt.votersForRoundCache = make(map[basics.Round]*ledgercore.VotersForRound)
+	vt.onlineAccountsFetcher = fetcher
 
-	latest := l.Latest()
-	hdr, err := l.BlockHdr(latest)
+	hdr, err := l.BlockHdr(latestDbRound)
 	if err != nil {
 		return err
 	}
 	proto := config.Consensus[hdr.CurrentProtocol]
 
-	if proto.CompactCertRounds == 0 || hdr.CompactCert[protocol.CompactCertBasic].CompactCertNextRound == 0 {
+	if proto.StateProofInterval == 0 || hdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound == 0 {
 		// Disabled, nothing to load.
 		return nil
 	}
 
-	startR := votersRoundForCertRound(hdr.CompactCert[protocol.CompactCertBasic].CompactCertNextRound, proto)
+	startR := votersRoundForStateProofRound(hdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound, proto)
 
 	// Sanity check: we should never underflow or even reach 0.
 	if startR == 0 {
 		return fmt.Errorf("votersTracker: underflow: %d - %d - %d = %d",
-			hdr.CompactCert[protocol.CompactCertBasic].CompactCertNextRound, proto.CompactCertRounds, proto.CompactCertVotersLookback, startR)
+			hdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound, proto.StateProofInterval, proto.StateProofVotersLookback, startR)
 	}
 
-	for r := startR; r <= latest; r += basics.Round(proto.CompactCertRounds) {
+	for r := startR; r <= latestDbRound; r += basics.Round(proto.StateProofInterval) {
 		hdr, err = l.BlockHdr(r)
 		if err != nil {
-			return err
+			vt.l.trackerLog().Errorf("votersTracker: loadFromDisk: cannot load tree for round %v, err : %v", r, err)
+			continue
```

Comment:
would this ever happen? why continue if you won't have the SP state you need? also the log message should mention failing to load the block header for this round.

Reply:
resolved in https://github.com/algorand/go-algorand/pull/4369